feat(settings): improve map pin ux

oktomus opened the pull request:
PR Checklist

- [x] - Commit [messages are descriptive](https://github.com/ONEARMY/community-platform/blob/master/CONTRIBUTING.md#--commit-style-guide), it will be used in our [Release Notes](https://github.com/ONEARMY/community-platform/releases/)

PR Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Developer experience (improves developer workflows for contributing to the project)

## Description

This is a first PR to improve the user experience when adding/editing/removing a map pin. 
I decided to split the work in multiple PRs because I might have done things that don't match with the code guidelines, and the figma design suggested in #2999  collides with the current implementation.

What this PR does
- Lock editing of the pin behind a button, so that you don't accidentally edit your pin
- Add a map image preview when you don't have a pin
- Show the pin in read only when you open your profile (`MapWithDraggablePin` becomes `MapBin` with a `draggable` property)

What I want to do next
- Noticed saving sometimes fail with no error when editing the pin, I assume it's not related to this PR but needs testing
- The readonly preview of your pin will be like the Maps page of the website, showing a badge, description and image

@thisismattia, in the [figma design](https://www.figma.com/file/N1O9TzrKijPaZfYXqBboC8/Community-Platform?type=design&node-id=910-3678&mode=design&t=rW80ghRehfvNryMR-0), I see a `Save map pin` button, but it kinda collides with the UI flow of `Save profile`. I see two ways going forward:
- Keep the current flow, user has to press `Save profile`, no `Save map pin` button
- Add a save button on the map + add a confirmation popup for `Remove pin`
Any thoughts?

## Git Issues

Closes #2999

## Screenshots/Videos

Here is a [loom ](https://www.loom.com/share/88404a9f07cf4b809eb1d23f01a4464b)showcasing the new UX 

---

## What happens next?

Thanks for the contribution! We try to make sure all PRs are reviewed ahead of our monthly maintainers call (first Monday of the month)

If the PR is working as intended it'll be merged and included in the next platform release, if not changes will be requested and re-reviewed once updated.

If you need more immediate feedback you can try reaching out on Discord in the [Community Platform `development` channel](https://discord.com/channels/586676777334865928/938781727017558018).

benfurber commented on src/pages/UserSettings/content/formSections/MemberMapPin.section.tsx:
```diff
@@ -43,29 +46,54 @@ export class MemberMapPinSection extends React.Component<any, IState> {
     })
   }
 
+  sectionTitle(): string {
+    return this.state.hasMapPin && !this.state.showAddressEdit
+      ? headings.map.yourPinTitle
+      : headings.map.addPinTitle
```
If you can work out how to get the return onto a single line I'd appreciate it. Doing `const { addPinTitle, yourPinTitle } = headings` would definitely be a good start.

benfurber commented on src/pages/UserSettings/content/formSections/MemberMapPin.section.tsx:
```diff
@@ -43,29 +46,54 @@ export class MemberMapPinSection extends React.Component<any, IState> {
+                justifyContent: 'space-between',
+              }}
+            >
+              <Text mt={4} mb={4} sx={{ display: 'block', flex: '2' }}>
```
Please do the margin setting in the sx.

benfurber commented on src/pages/UserSettings/content/formSections/MemberMapPin.section.tsx:
```diff
@@ -43,29 +46,54 @@ export class MemberMapPinSection extends React.Component<any, IState> {
+              <Text mt={4} mb={4} sx={{ display: 'block', flex: '2' }}>
+                {headings.map.description}
+              </Text>
+              <Box sx={{ flex: '1' }}>
```
Interesting, is adding the flex 1 needed? And if it is needed shouldn't it be a number rather than a string?

oktomus replied:
Without flex 1, it looks like this 
![image](https://github.com/ONEARMY/community-platform/assets/4656466/bae638b7-d03b-42e3-aace-a0adf037084a)

benfurber commented on src/pages/UserSettings/content/formSections/MemberMapPin.section.tsx:
```diff
@@ -81,7 +109,58 @@ export class MemberMapPinSection extends React.Component<any, IState> {
+                showCharacterCount
+                validate={required}
+                validateFields={[]}
+                value={this.injected.userStore.user?.mapPinDescription}
```
Once you rebase you'll see this component is now a hook so you'll be accessing the userstore differently.

benfurber reviewed (changes requested):
This is a really like improvement to the app, thanks for picking it up @oktomus. A bunch of comments added and a couple more thoughts below.

I think it's worth pulling out DraggableMarker into it's own file with appropriate storybook/tests please.

There's a bunch more destructuring of useful objects about the place so things like `this.state.hasMapPin` don't have to be called all the time.

Please shout if you need me to clarify anything.

oktomus commented on packages/components/src/MapWithDraggablePin/MapPin.tsx:
```diff
@@ -74,48 +79,50 @@ export const MapWithDraggablePin = (props: Props) => {
-                  },
-                )
+        {props.draggable && (
+          <Flex style={{ width: '280px' }}>
```
This is not something I introduced I believe. Here is how it looks on mobile: 

![image](https://github.com/ONEARMY/community-platform/assets/4656466/abbf26fc-dd5d-475a-a04a-bd9b516f3539)

benfurber replied:
Oh I see, it looks new but it's moved from `MapWithDraggablePin.tsx`. 👍

oktomus commented:
Thanks for the review @benfurber! I addressed all the comments. For only two of them I didn't change anything but I commented. 
I introduced a stories file for the MapPin but I'm not sure what is the purpose of it, tell me if I need to change things.

benfurber commented:
@thisismattia I'm happy with this from a code prospective, does it behave on preview as you'd expect?

oktomus commented:
Were you guys able to have a look?

davehakkens commented:
hey sorry for the extremely slow reply @oktomus and thanks for submitting such a complete PR. (with loom 🙏)

It works really nice. Behaviour like this makes sense and doesnt need to be aligned with the Figma so you can keep the flow as it is.

I do have one comment. It looks like you only implemented this for the `member` profiles not the other profile types
You can see them when you toggle here:

<img width="809" alt="afbeelding" src="https://github.com/ONEARMY/community-platform/assets/13672737/842c5fba-42f8-4e5a-98f8-4459db6ec7cf">

Could you implement it for the other profile types as well?

oktomus commented:
Hey @davehakkens, sure I will have a look :)

oktomus commented:
@davehakkens @benfurber Changes have been made, we now have only one component for the map pin section that is shared between members and workspace. 

I noticed that the stars warning for workspace was always there, I left it as it but is that normal? 
![image](https://github.com/ONEARMY/community-platform/assets/4656466/17facf30-f5d0-4b38-b650-1bd00f5919a1)




Sorry for the delay!

P.S: will there be a dev mission at Project Kamp next year? 🙏

oktomus commented:
A bug I found with the map pin is that if you place your pin somewhere in the world where there is no street close-by, saving will fail because we don't have an address, but no warning/error message is being displayed.  It's something I could take a look at in another PR. 

Should I create an issue for it?

oktomus commented:
@benfurber Do you know what can I do about the lock file being modified? It's in the CI failures 
![image](https://github.com/ONEARMY/community-platform/assets/4656466/7bd67ec0-d5b2-4bd5-b79d-60705e7ac24f)

benfurber commented:
@oktomus! Thank you so much for coming back to this after we basically blanked you for a bit! 

I think the CI issue relates to problem we had a while back so I've merged in master and I think that'll fix it. 

Yes please to raising the issue you've found.

Looks like there's a linting issue though? `yarn format` should sort that right out.

benfurber reviewed (changes requested):
Annoyingly I thought we where there, but there's a cypress test failing! I tried to quickly add `    cy.get('[data-cy=add-a-map-pin]').click()` in the right spot @oktomus. But there's more to it than that. Looks like there's some wierd state stuff going on with the form. I'll have to leave it with you.

benfurber commented:
This happened to me on another form change. This might help: https://github.com/ONEARMY/community-platform/pull/3214

oktomus commented:
Ok I will take a look :)

oktomus commented:
@benfurber I took a look but I'm not sure how to fix it. Checkout this video for explanations: https://www.loom.com/share/a298aec7a7af4a2ebfeb68a3d3433a6c

Something is going off when calling setState

pizzaisdavid commented:
Hope this isn't rude but I pulled the branch locally and tried to do some stuff.

## Summary
I have fixed the merge conflicts and unit tests. See the [code comparison here](https://github.com/ONEARMY/community-platform/compare/master...pizzaisdavid:community-platform:feat/oktomus-improve-map-pin-editor-ux?expand=1). (The branch on my fork is called `feat/oktomus-improve-map-pin-editor-ux`, [link](https://github.com/pizzaisdavid/community-platform/tree/feat/oktomus-improve-map-pin-editor-ux))

In the current state, there is a separate tab for the map, but switching between the tabs clears un-saved changes. And because there is no save button on the map tab, it is not possible to save map changes. Some e2e tests fail for that reason.

I'm not sure how would be best to precede: make the un-saved data stay between tab changes or putting the map on the original page. (I am unsure if the tab was introduced to resolve the state bug or if it was a user design decision. I checked the figma in the ticket and didn't see a separate tab there.)

I didn't fix the e2e tests because I wasn't sure which direction things will go, but there is a work-in-progress commit on my branch.

@oktomus if you would like to pair on this (either the state bug or making the un-saved data keep between tab changes) you can ping me.

I suppose I could pop the commit for making it a tab and investigate the bug myself. Or also to clarify, @oktomus did you ever figure the bug out?

## Details

Getting the branch on my local:
```
git remote add oktomus git@github.com:oktomus/community-platform.git
git fetch oktomus feat/improve-map-pin-editor-ux
git checkout feat/improve-map-pin-editor-ux
```
running on my localhost:
```
sudo sysctl fs.inotify.max_user_watches=131070
yarn
yarn start
```

fixing the merge conflicts:
(I already have the latest ONEARMY `master` because I forked ~10 minutes ago.)
```
git rebase master
```
There is a file, `MapWithDraggablePin.tsx` that was deleted on master but this merge request edits it.
To fix: delete it.

Then:
```
git rebase --continue
```
There is another merge conflict, in the test file `src/pages/UserSettings/UserSettings.test.tsx`

Relevant part on master:
```
      <ThemeProvider theme={Theme}>
        <MemoryRouter
          initialEntries={[routerInitialEntry ? routerInitialEntry : '']}
        >
          <SettingsPage adminEditableUserId={isAdmin ? user._id : null} />
        </MemoryRouter>
      </ThemeProvider>
```

Relevant part on this branch:
```
      <ThemeProvider theme={Theme}>
        <MemoryRouter>
          <SettingsForm adminEditableUserId={isAdmin ? user._id : null} />
        </MemoryRouter>
      </ThemeProvider>
```
There is a change in the things passed to the MemoryRouter, but also, there is a difference of SettingsForm and SettingsPage. The correct thing to (probably) fix the test seems to be to change it to `SettingsContainer` as that was something introduced in this merge request to contain both tabs (I'd be in favor of just calling that `SettingsPage` but I am trying to make the minimal changes here.)

After making that change and updating the import. Running the unit tests (`yarn test:unit`), a single test fails (`UserSettings › map pin › displays moderation comments to user`) makes sense generally, we now need to click the map tab. (technically, there are two tests relating to the map pin but one passes anyways because it is checking something doesn't exist.)

Seems like unit tests use `userEvent` to simulate user interaction. We need to click the map tab, add the following to both tests after the `act` block:
```
const mapButton = wrapper.getByText('Map')
await waitFor(() => mapButton.click())
```
Unit tests now pass, nice.

Time to check e2e tests: `yarn test`

Some fail. Add: `data-cy="map-tab"` to the `<Tab>` element.

Authors note: at this point I found the bug mentioned in the summary and wasn't sure how to continue.

oktomus commented:
Hi there, thanks for having a look @pizzaisdavid! 

The tab is something we added together with @benfurber, but we haven't finished it and I didn't take the time to work on it since. Originally, there was no tab, and one test was failing. 

Because a big redesign-refactor into tabs was planned for the profile page, we thought it was better to not fix the test and just start the redesign-refactor.

The best I think would be to have the map tab working. The tab is not a work-around, eventually there would be a tab for each section. I don't remember where but there is another place on the website where tabs are used, and the goal is to do the same in the profile page (you can probably find where by looking for Tab component usages).

However, maybe it's easier for you to fix the test instead of working on the tabs. In that case feel free to remove the commit and fix the test, maybe my last comments about it can give you some hints on what is wrong. But, you would be fixing something that would probably disappear some time soon.

Unfortunately I don't have time to sync at the moment, but I will try my best to follow this conversation and answer any question you might have!

pizzaisdavid commented:
Thanks @oktomus for the information. It was helpful. 😃 

Since a redesign is desired, then I agree, the best way forward is to continue with the tab.

@oktomus @benfurber Do you remember/know if there was a UI mock-up? My main question being if there should be a separate save button for the map pin. (I checked the Figma and didn't see anything for that https://www.figma.com/design/N1O9TzrKijPaZfYXqBboC8/Community-Platform)

@benfurber if there isn't anything, can I just make something up?

My proposal would be:
- Switching tabs doesn't clear the data between them.
- Each tab has its own save button for now.
- Ideally, when the user goes to close the tab, they'd get a warning if they have unsaved changes (unsure if like that now.)

Also, I don't promise that I'd get to this. 😜

davehakkens commented:
FYI @dalibormrska is looking into the redesign of this

pizzaisdavid commented:
If there isn't an understanding yet of how the design should look/behave, then I would propose to remove the "awaiting author" label and add "Blocked" and/or "Under Discussion" labels.

benfurber commented:
As we're on a kick to get all older PRs closed I had a look at this. I removed the new form design commit. I started to look into the bug that caused an issue a while back and, eerrr, it seems to have resolved itself!

pizzaisdavid commented:
Code review: Besides stuff like using shorthand variable names, evt, lng, lat. looks okay to me. Good job for giving this a push :slightly_smiling_face:

oktomus commented:
Well done for fixing the bug somehow :D

benfurber commented:
@all-contributors add @oktomus for code
